solution-jivanov/models/users.js

```diff
@@ -0,0 +1,283 @@
+// users.js
+// users model class
+
+const uuidv4 = require('uuid/v4');
+const validate = require('validate.js');
+
+const Model = require('../lib/model.js');
+
+const ID_VALIDATION_CONSTRAINTS = {
+    presence: true, // it is mandatory
+    type: 'string', // ID must be of type string, since it is an UUID
+    length: {
+        is: 36, // and must be exactly 36 chars too
+        message: 'The record ID must be exactly 36 chars long'
+    },
+};
+const CREATE_VALIDATION_CONSTRAINTS = {
+    name: {
+        presence: true, // only name is mandatory
+        length: {
+            minimum: 3, // and must be at least 3 chars long
+            message: 'field is mandatory and must be at least 3 chars long'
+        },
+    },
+    dob: {
+        presence: false,
+    },
+    address: {
+        presence: false,
+        length: {
+            minimum: 2, // and must be at least 2 chars long
+            message: 'must be at least 2 chars long'
+        },
+    },
+    description: {
+        presence: false,
+        length: {
+            minimum: 2, // and must be at least 2 chars long
+            message: 'must be at least 2 chars long'
+        },
+    },
+};
+const UPDATE_VALIDATION_CONSTRAINTS = {
+    name: {
+        presence: false, // only name is mandatory
+        length: {
+            minimum: 3, // and must be at least 3 chars long
+            message: 'must be at least 3 chars long'
+        },
+    },
+    dob: {
+        presence: false,
+    },
+    address: {
+        presence: false,
+        length: {
+            minimum: 2, // and must be at least 2 chars long
+            message: 'must be at least 2 chars long'
+        },
+    },
+    description: {
+        presence: false,
+        length: {
+            minimum: 2, // and must be at least 2 chars long
+            message: 'must be at least 2 chars long'
+        },
+    },
+};
+const VALIDATION_OPTIONS = {
+    format: 'flat', // return "flat" message format i.e. a single string
+};
+
+const USERS_TABLE_NAME = process.env.USERS_TABLE_NAME;
+
+// a map to translate input params to DynamoDB ones if needed
+const dynamoDbKeys = new Map ([
+    ['name', '#userName'],
+]);
+
+
+class Users extends Model {

```

Comment:
It seems like this Users model class, specifically, is responsible for input validation, data access layer communication details, and generic abstraction of user methods.

I would like to understand your approach on scope-limiting OOP for Models.

Reply:
Certainly, let me try to elaborate please. My approach here is KISS. 
As per wikipedia a model is "The central component of the pattern. It is the application's dynamic data structure, independent of the user interface.[5] It directly manages the data, logic and rules of the application." [Reference here](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller)

Everything that directly works with the Users data would go there. 
Anything else would go to a different model - for example if the users have roles which are defined (and perhaps stored) somewhere else, then that would be a separate model (UserRoles)

I would not try to force scope-limiting a model unless it becomes too big/unreadable. In the case of this particular model, we could split it in like "UserValidations", "UserDatabase" and so on.

But again, in general, I find it much easier, much more efficient to work with a single file (assuming it is not too big of course), it is just so convenient to have everything under your finger tips...